File: genetic-packer/src/main/java/genetic/packer/fx/calculation/impl/BoundingSizeCalculatorCreator.java
package genetic.packer.fx.calculation.impl;

import genetic.packer.fx.calculation.BoundingSizeCalculator;
import javafx.geometry.Bounds;
import javaslang.collection.Traversable;
import org.springframework.stereotype.Component;

/**
 * @author piotr.larysz
 */
@Component
public class BoundingSizeCalculatorCreator implements BoundingSizeCalculator.Creator {

    @Override
    public BoundingSizeCalculator from(Traversable<Bounds> allBoxesBounds) {
        return tuple ->  tuple.transform((minimumProperty, maximumProperty) -> {
            final double max = allBoxesBounds.map(maximumProperty).max().getOrElse(0.0);//todo maybe throw some exception?
            final double min = allBoxesBounds.map(minimumProperty).min().getOrElse(0.0);
            return max - min;
        });
    }
}

File: genetic-packer/src/main/java/genetic/packer/fx/calculation/impl/BoundingSizeCalculatorCreator.kt
package genetic.packer.fx.calculation.impl

import genetic.packer.fx.calculation.BoundingSizeCalculator
import genetic.packer.fx.specification.BoundsGetter
import javafx.geometry.Bounds
import org.springframework.stereotype.Component

@Component
class BoundingSizeCalculatorCreator : BoundingSizeCalculator.Creator {

  override fun from(bounds: Sequence<Bounds>) =
      object : BoundingSizeCalculator {

        override fun invoke(pair: Pair<BoundsGetter, BoundsGetter>): Double {
          pair.let { (minimumProperty, maximumProperty) ->
            val max = bounds.map { maximumProperty.apply(it) }
                .max() ?: 0.0
            val min = bounds.map { minimumProperty.apply(it) }
                .min() ?: 0.0
            return max - min
          }
        }
      }
}